src/main/scala/net/snowflake/spark/snowflake/io/CloudStorageOperations.scala

```diff
@@ -563,188 +559,158 @@ sealed trait CloudStorage {
-        s"""${SnowflakeResultSetRDD.WORKER_LOG_PREFIX}:
-           | Finish writing partition ID:$partitionID:
-           | without cache data. Total process time is
-           | ${Utils.getTimeString(endTime - startConvertTime)}
-           | write row count is $rowCount.
-           | Uncompressed data size is ${Utils.getSizeString(dataSize)}.
-           | (Data size = 0 means file is empty, no file is uploaded)
-           |""".stripMargin.filter(_ >= ' '))
-    } else if (readData.isDefined && readData.get.nonEmpty) {
-      // Upload the file with retry and backoff.
-      // default maxRetryCount is 10 which is configurable.
-      var retryCount = 0
-      var error: Option[Exception] = None
-      var uploadDone = false
-      var startUploadTime: Long = 0
-      val data = readData.get
-      dataSize = data.size
-      do {
-        try {
-          startUploadTime = System.currentTimeMillis()
-
-          if (storageInfo.isDefined) {
-            // Update data with StorageInfo
-            val uploadStream =
-              createUploadStream(fileName, Some(directory), compress, storageInfo.get)
-            uploadStream.write(data)
-            uploadStream.close()
-          } else if (fileTransferMetadata.isDefined) {
-            // Set up proxy info if it is configured.
-            val proxyProperties = new Properties()
-            proxyInfo match {
-              case Some(proxyInfoValue) =>
-                proxyInfoValue.setProxyForJDBC(proxyProperties)
-              case None =>
-            }
-
-            // Update data with FileTransferMetadata
-            val inStream = new ByteArrayInputStream(data)
-            SnowflakeFileTransferAgent.uploadWithoutConnection(
-              SnowflakeFileTransferConfig.Builder.newInstance()
-                .setSnowflakeFileTransferMetadata(fileTransferMetadata.get)
-                .setUploadStream(inStream)
-                .setRequireCompress(compress)
-                .setOcspMode(OCSPMode.FAIL_OPEN)
-                .setProxyProperties(proxyProperties)
-                .build())
-          }
-          val endUploadTime = System.currentTimeMillis()
-
-          TestHook.raiseExceptionIfTestFlagEnabled(
-            TestHookFlag.TH_GCS_UPLOAD_RAISE_EXCEPTION,
-            "Negative test to raise error when uploading data to GCS"
-          )
+    try {
+      // Read data and upload to cloud storage
+      doUploadPartition(rows, format, compress, directory, partitionID,
+                        storageInfo, fileTransferMetadata)
+    } catch {
+      // Hit exception when uploading the file
+      case e: Exception => {
+        val attemptNumber = TaskContext.get().attemptNumber()
+        val sleepTime = retrySleepTimeInMS(attemptNumber + 1)
+        // Sleep exponential time based on the attempt number.
+        Thread.sleep(sleepTime)
+
+        val stringWriter = new StringWriter
+        e.printStackTrace(new PrintWriter(stringWriter))
+        val errmsg =
```

Comment:
Let's log befgore we sleep and let's make sure the sleep time, attempt number and task ids are added  as part of the logs.

Reply:
Sure.